sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,578 @@
+      val actualNumDigits: Int =
+        if (i < parsedDigitGroupSizes.length) {
+          parsedDigitGroupSizes(i)
+        } else {
+          0
+        }
+      expectedToken match {
+        case ExactlyAsManyDigits(expectedNumDigits)
+          if actualNumDigits != expectedNumDigits =>
+          // The input contained more or fewer digits than required.
+          return None
+        case AtMostAsManyDigits(expectedMaxDigits)
+          if actualNumDigits > expectedMaxDigits =>
+          // The input contained more digits than allowed.
+          return None
+        case _ =>
+      }
+    }
+    Some(inputIndex)
+  }
+
+  /**
+   * Returns true if the given character matches a digit (0-9) or a comma, updating fields of
+   * this class related to parsing during the process.
+   */
+  private def matchesDigitOrComma(char: Char, reachedDecimalPoint: Boolean): Boolean = {
+    char match {
+      case char if char.isWhitespace =>
+        // Ignore whitespace and keep advancing through the input string.
+        true
+      case char if char >= ZERO_DIGIT && char <= NINE_DIGIT =>
+        numDigitsInCurrentGroup += 1
+        // Append each group of input digits to the appropriate before/afterDecimalPoint
+        // string for later use in constructing the result Decimal value.
+        if (reachedDecimalPoint) {
+          afterDecimalPoint.append(char)
+        } else {
+          beforeDecimalPoint.append(char)
+        }
+        true
+      case COMMA_SIGN =>
+        parsedDigitGroupSizes.prepend(numDigitsInCurrentGroup)
+        numDigitsInCurrentGroup = 0
+        true
+      case _ =>
+        parsedDigitGroupSizes.prepend(numDigitsInCurrentGroup)
+        false
+    }
+  }
+
+  /**
+   * This method executes when the input string fails to match the format string. It throws an
+   * exception if indicated on construction of this class, or returns NULL otherwise.
+   */
+  private def formatMatchFailure(input: UTF8String, originNumberFormat: String): Decimal = {
+    if (errorOnFail) {
+      throw QueryExecutionErrors.invalidNumberFormatError(input, originNumberFormat)
+    }
+    null
+  }
+
+  /**
+   * Computes the final Decimal value from the beforeDecimalPoint and afterDecimalPoint fields of
+   * this class, as a result of parsing.
+   *
+   * @param negateResult whether the input string specified to negate the result
+   * @return a Decimal value with the value indicated by the input string and the precision and
+   *         scale indicated by the format string
+   */
+  private def getDecimal(negateResult: Boolean): Decimal = {
+    // Append zeros to the afterDecimalPoint until it comprises the same number of digits as the
+    // scale. This is necessary because we must determine the scale from the format string alone but
+    // each input string may include a variable number of digits after the decimal point.
+    val extraZeros = "0" * (scale - afterDecimalPoint.length)
```

Comment:
do we still need to do it? `Decimal(javaDecimal, precision, scale)` should take care of the scale already.

Reply:
Yes apparently :) I tried without this part and it failed because the `scale` is assigned at analysis time based on the format string. So if e.g. the format string is `999.99` then the precision is 5 and the scale is 2, but if the input string is `123.4` then the `afterDecimalPoint.length` is 1, and thus creating the final `Decimal` with `scale` set to 2 results in incorrect results.